### PR TITLE
feat: add CONTRIBUTING guide

### LINKED ISSUE #122: Add CONTRIBUTING guide to repo following practice at jenkins-infra/jenkins.io repo
### Describe your use-case which is not covered by existing documentation.

As we will be replacing the existing https://www.jenkins.io/ website with the docs in this repo, we would like to add a similar "CONTRIBUTING.adoc" file serving as a guide for future contributions as currently exists at https://github.com/jenkins-infra/jenkins.io/blob/master/CONTRIBUTING.adoc. 

Please note that some copy-editing and modifications may be needed to complete the task to satisfaction. 

### Reference any relevant documentation, other materials or issues/pull requests that can be used for inspiration.

_No response_

### PR DESCRIPTION
**Pull request by adripo**
**Description:**

This pull request introduces a CONTRIBUTING guide to the repository. The guide outlines the steps for contributing. This addition aims to enhance collaboration and streamline the contribution process for new and existing contributors.

**Changes Made:**

- Created a `CONTRIBUTING.adoc` file.
- Improved clarity and organization for contributors.

**Related Issue:**
- This PR closes #122

### CODE REVIEW

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:1]:
```suggestion
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:2]:
```suggestion
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:3]:
```suggestion
```

File: CONTRIBUTING.adoc
Comment (krisstern) [reply CONTRIBUTING.adoc:1]: The first few lines are no longer needed as we are no longer using the Ruby-based framework Awestruct for the build tool in this repo.

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:4]:
```suggestion
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:9]:
```suggestion
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:10]:
```suggestion
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:206]:
```suggestion
[[why-asciidoc]]
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:23]:
```suggestion
```

Comment (krisstern) [reply CONTRIBUTING.adoc:23]: We do not have Gitpod configured for this revamped version of the website yet.

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:24]:
```suggestion
```

**Review by krisstern — CHANGES_REQUESTED** [review]

**Comment by krisstern** [comment]
Thanks for your PR @adripo! There are some changes needed before this PR can be approved and accepted for Hacktoberfest 2024.

**Comment by adripo** [comment]
Thanks for the suggestions @krisstern. I fixed everything.

**Comment by krisstern** [comment]
Hi @adripo I noticed you used your own `main` branch instead of a feature branch to submit code changes to this repo. This is not recommended as I cannot review your code locally normally. For more info please see: https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:48]:
```suggestion
   - link: link:./docs/README.adoc[docs]
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:53]:
```suggestion
   - link: link:./site/README.md[site]
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:301]:
```suggestion
```

Comment (krisstern) [reply CONTRIBUTING.adoc:301]: We do not have this `./.github/CODEOWNERS` file set up yet.

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:238]:
```suggestion
```

Comment (krisstern) [reply CONTRIBUTING.adoc:238]: We do not yet have this "./.github/CODEOWNERS" file set up yet.

**Review by krisstern — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.adoc
Comment (adripo) [reply CONTRIBUTING.adoc:48]: The link to the readme is 2 rows below. This is just the link to the whole section

https://github.com/jenkins-infra/docs.jenkins.io/pull/141/files/9d61d1f3db63475c45ae6ae5f63751f3c392af8f#diff-cf6df8a86bba8fb7539f0419c4cc2c7033ca75beb25a7979548068b5f01d439fR50

Comment (adripo) [reply CONTRIBUTING.adoc:48]: let me know if you prefer to link the readme here too

File: CONTRIBUTING.adoc
Comment (adripo) [reply CONTRIBUTING.adoc:53]: Same here, the link to the readme is 2 rows below

**Comment by adripo** [comment]
> Hi @adripo I noticed you used your own `main` branch instead of a feature branch to submit code changes to this repo. This is not recommended as I cannot review your code locally normally. For more info please see: https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

Hi @krisstern I understand your point about using a feature branch. I can create a new branch from my `main` if that helps, but I don't believe it directly addresses the issue.

To fully adhere to the _Git feature branch workflow_, I would need write access to the `jenkins-infra` project to create a feature branch directly within the repo. In my case, using a fork means that whether I use the `main` branch or a feature branch in my fork, the outcome is the same since I’m working from my own repository.

However, I will try to follow that practice for the next contributions.

File: CONTRIBUTING.adoc
Comment (krisstern) [reply CONTRIBUTING.adoc:48]: Your original link is broken, and we don't want a slash in the text for the link. I have just edited the suggestion. Please commit the change.

File: CONTRIBUTING.adoc
Comment (krisstern) [reply CONTRIBUTING.adoc:53]: Same here, your link is broken and we don't want a slash in the text. I have just updated the suggestion. Please commit the change.

**Comment by krisstern** [comment]
> > Hi @adripo I noticed you used your own `main` branch instead of a feature branch to submit code changes to this repo. This is not recommended as I cannot review your code locally normally. For more info please see: https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow
> 
> Hi @krisstern I understand your point about using a feature branch. I can create a new branch from my `main` if that helps, but I don't believe it directly addresses the issue.
> 
> To fully adhere to the _Git feature branch workflow_, I would need write access to the `jenkins-infra` project to create a feature branch directly within the repo. In my case, using a fork means that whether I use the `main` branch or a feature branch in my fork, the outcome is the same since I’m working from my own repository.
> 
> However, I will try to follow that practice for the next contributions.

Actually what you are doing means I cannot push code to your branch as a maintainer if you are not aware.

**Comment by krisstern** [comment]
Also, by using a feature branch I mean any other branch that is not named `main`.

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:50]:
```suggestion
   - You can find more information in the link:./docs/README.adoc[Antora README.adoc].
```

File: CONTRIBUTING.adoc
Comment (krisstern) [inline CONTRIBUTING.adoc:55]:
```suggestion
   - For more details, you can refer to the link:./site/README.md[Gatsby README.md].
```

**Review by krisstern — CHANGES_REQUESTED** [review]
The original links provided are not the correct ones.

**Comment by adripo** [comment]
Hi @krisstern, I appreciate your guidance regarding the use of feature branches. I want to clarify that I'm fully aware of what a feature branch is and understand the purpose of using branches other than `main`.

The “Allow edits by maintainers” option is checked for my PR, which should enable maintainers to push commits directly to my branch. Additionally, my branch protection rules do not restrict changes to the `main` branch.

If there's specific documentation or a resource that explains the behavior you're mentioning, I would appreciate you sharing that with me. It would help me understand better and improve my future contributions.

Regardless, I've applied your edits, and I see that the PR is ready to be merged.

**Review by krisstern — APPROVED** [review]
Thanks @adripo! LGTM

**Comment by krisstern** [comment]
> Hi @krisstern, I appreciate your guidance regarding the use of feature branches. I want to clarify that I'm fully aware of what a feature branch is and understand the purpose of using branches other than `main`.
> 
> The “Allow edits by maintainers” option is checked for my PR, which should enable maintainers to push commits directly to my branch. Additionally, my branch protection rules do not restrict changes to the `main` branch.
> 
> If there's specific documentation or a resource that explains the behavior you're mentioning, I would appreciate you sharing that with me. It would help me understand better and improve my future contributions.
> 
> Regardless, I've applied your edits, and I see that the PR is ready to be merged.

Acoording to Git rules, you can only push a branch with the same name from local to remote, so that if you name your branch `main` and since I cannot and will not name my fetched branch from your repo `main`, as it would cause problems to my local setup, it is quite impossible (if you have enabled maintainers to edit your branch) for me to push changes to your branch while your are calling your feature branch `main`.

**Comment by adripo** [comment]
@krisstern Thank you for the clarification, I now understand that the issue stems from your local setup. However, you can still check out the PR locally with a custom branch name using the following command:

```
git fetch origin pull/$ID/head:$BRANCH_NAME && git checkout $BRANCH_NAME
```

For example, you could use something like `contributing` to make it clear. This process is documented here: [Checking out pull requests locally](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally)

I generally do follow feature branch practices, but I started this task from the online editor, which automatically kept the `main` branch. 

Anyway, it was a pleasure to work on this task, and I look forward to future contributions.

**Comment by krisstern** [comment]
Hi @adripo,
It is really some very common and basic Git knowledge that is a given for contributing to open source. You seem to be having some problem understanding my point: I already did check out your branch with a new name that is not `main`, but I cannot push to it because I am not going to change my `main` branch to your `main` branch, and you are the one using the `main` brancch.

**Comment by adripo** [comment]
@krisstern I see there has been a miscommunication. My intention was to offer a solution for local setup issues while acknowledging that using feature branches is the best practice, which I usually follow. If you follow my suggestions above you'll be able to push even if the local name is different with a command like
```
git push origin $BRANCH_NAME:pull/$ID/head
```

As already said, I’ll make sure to adhere to the preferred workflow in future contributions. Thanks for your time.

**Comment by krisstern** [comment]
Thanks but no thanks

**Comment by krisstern** [comment]
You original told me to do `git fetch origin pull/$ID/head:$BRANCH_NAME && git checkout $BRANCH_NAME` in an earlier post and then you change it to `git push origin $BRANCH_NAME:pull/$ID/head` in a more recent post. Actually you seem kind of confused. What are you trying to achieve here?

**Comment by adripo** [comment]
@krisstern I’m quite comfortable managing pull requests and I am more than happy to help you understand better. Let me clarify:

The first command I shared:

```bash
git fetch origin pull/$ID/head:$BRANCH_NAME && git checkout $BRANCH_NAME
```
was intended for **checking out the pull request** locally, so you can review or make changes to the code.

The second command:

```bash
git push origin $BRANCH_NAME:pull/$ID/head
```
was meant for **pushing any changes back to the pull request** after you've made your modifications locally. This command specifically updates the PR branch by pushing changes from your local branch.

So, the two commands serve different purposes: one for fetching and checking out the PR locally, and the other for pushing changes back to the PR.

I hope this clears up any confusion.

**Comment by krisstern** [comment]
Hi @adripo please be aware we do have our code of conduct at https://www.jenkins.io/project/conduct/. Your behavior seems to be violating our code of conduct. If you offend again I will need to report you to the persons concerned (i.e., the Jenkins Board) within the community.

**Comment by krisstern** [comment]
Hi @adripo, is there a reason why you are so very keen to give unwelcomed as well as unsolicited advice that may or may not work?

**Comment by adripo** [comment]
Hi @krisstern, thank you for bringing this to my attention. I want to clarify that my intentions have always been to bring up a constructive dialogue and assist with any local management issues you might have encountered.

I have not offered unsolicited advice; instead, I aimed to explain the reasoning behind my choices in the development process and provide solutions based on the context of our discussion. Initially, when you mentioned concerns about using the `main` branch in my fork, I responded positively, indicating I would be happy to create a new branch if it would help.
I also clarified that the Git workflow documentation you referenced does not relate to the branch I used in my fork.

I would appreciate it if you could specify what aspects of my behavior you believe violate the code of conduct, as I strive to maintain professionalism and respect on this platform.

Thank you for your understanding, and I hope we can move forward positively.

**Comment by krisstern** [comment]
@adripo I no longer wish to have any dialog or interaction with you, as your actions shows no respect for other people at all, despite pretending to be otherwise. Having no respect for other people's wishes or feelings is already a violation of our code of conduct. I will cease all future communication with you from this point onwards. 

The so-called instructions you offered is simply incorrect and does not help anyone on anything. (TBH I really need to create a remote to your fork first before I can push to your `main`, I can sacrifice my `main` branch temporarily to make this happen, and I just need to regnerate my `main`, but like I said I will not and in fact need not do that. So you see you are missing some crucial knowledge you have no idea about it seems.)

**Comment by gounthar** [comment]
> Hi @krisstern, I appreciate your guidance regarding the use of feature branches. I want to clarify that I'm fully aware of what a feature branch is and understand the purpose of using branches other than `main`.
> 
> The “Allow edits by maintainers” option is checked for my PR, which should enable maintainers to push commits directly to my branch. Additionally, my branch protection rules do not restrict changes to the `main` branch.
> 
> If there's specific documentation or a resource that explains the behavior you're mentioning, I would appreciate you sharing that with me. It would help me understand better and improve my future contributions.
> 
> Regardless, I've applied your edits, and I see that the PR is ready to be merged.

Quoting https://hynek.me/articles/pull-requests-branch/#:~:text=Here%20are%20the%20three%20main,annoying%20faster%20than%20one%20think.:

> This may sound like pedantry or forcing the Right Way™ to do things on you. It’s not – I’m trying to save you (and the benefactor of your contribution) pain.
> 
> Here are the three main reasons:
> 
> You can only work on one pull request at a time, unless you can [stack](https://newsletter.pragmaticengineer.com/p/stacked-diffs) the new pull request on top of the old one. You can have just one fork and there’s just one main branch.
> 
> This can get annoying faster than one think. For example, if you need to fix something small before doing the main work.
> 
> If the target repo merges your pull request (yay!), you can’t pull from it anymore because you’ve got conflicting commits (boo!).
> 
> How complicated it is to recover from this situation depends on the target repo’s merge strategy. In the worst case (you had more than one commit in the pull request and the maintainer does a squash-and-merge for a clean history), you’re looking at a git reset --hard – or just nuking the fork and starting over.
> 
> If the target repo has branch protection on (many do), the [“Allow edits from maintainers.” checkbox](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) becomes meaningless and maintainers can’t push to your branch. Meaning: all changes have to go through you.

Had I seen this pull request earlier, I would have closed it.
We are thankful for your contributions, but we have rules. One of them is that we don't accept pull requests from the main branch of your fork.
It's good for the project and for your fork.

**Comment by gounthar** [comment]
# Community Guidelines and Code of Conduct Reminder

@adripo, I appreciate your contributions to the Jenkins project. However, I need to address some concerns about recent interactions within our community.

## Community Values

Jenkins is more than just a codebase - it's a vibrant community built on mutual respect and collaboration. Our strength comes from:

1. Open and constructive dialogue
2. Respect for differing viewpoints
3. Focus on collective growth

## Positive Behaviors We Encourage

Our [Code of Conduct](https://www.jenkins.io/project/conduct/) emphasizes behaviors that create a positive environment:

- Demonstrating empathy and kindness
- Respecting different opinions and experiences
- Exchanging constructive feedback professionally
- Taking responsibility for mistakes and learning from them
- Prioritizing community benefit over individual preferences

## Moving Forward

I encourage you to:
1. Review our [Code of Conduct](https://www.jenkins.io/project/conduct/)
2. Consider how your communication style impacts others
3. Frame future discussions in a more collaborative manner

Remember that technical excellence and respectful communication are equally important in our community.

We value your contributions and look forward to your continued participation in alignment with our community guidelines.

**Comment by krisstern** [comment]
To clarify, I never experienced any local setup issues. I do not have any clue where that idea is coming from.

**Comment by adripo** [comment]
Hi @gounthar, thank you for your input and for sharing the useful link, it's much appreciated. I completely agree with the best practice of not starting pull requests from the `main` branch, which is something I usually follow. 

In this case, the branch remained as `main` because I started working on the fork directly from GitHub's browser tools before checking out the repository locally and just later on saw my issue. That’s why I offered to create a separate branch and recreate the PR from a proper feature branch, but I didn’t receive confirmation to proceed, so I continued working on this one.

As I’ve mentioned in previous messages, I will make sure to adhere to this practice correctly in future contributions.

Regarding the community guidelines, I sincerely apologize if my behavior caused any issues. My goal has always been to respond politely and contribute useful information to the discussion, even when some replies were rather sharp. I hope we can move forward in a more collaborative way in the future.

Thanks again for your understanding, and I look forward to continued collaboration!

**Comment by gounthar** [comment]
Thank you for your thoughtful response and for explaining the circumstances regarding the branch situation. Your openness to feedback and willingness to adapt are exactly the kind of collaborative spirit we value in our community.

I appreciate your detailed explanation about the GitHub browser tools leading to the `main` branch usage. This kind of transparency helps everyone understand the situation better, and your proactive offer to recreate the PR from a feature branch demonstrates good intentions.

Your gracious response to the community guidelines reminder is commendable. We all learn and grow together, and your commitment to moving forward collaboratively is greatly appreciated.

Please continue contributing to Jenkins - your contributions are valuable, and your positive attitude toward community interaction makes our project stronger.

Looking forward to our continued collaboration!

**Comment by krisstern** [comment]
@adripo Apology accepted, but please do not offer advice or suggestion of any form to anyone in the Jenkins community you don't seem to have full command of, at least not yet, unless you are explicitly being asked for such advice or suggestion.

**Pull request merged**
